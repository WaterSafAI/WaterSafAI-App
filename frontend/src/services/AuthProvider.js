import React, {createContext, useContext, useMemo, useState} from 'react';
import auth from '@react-native-firebase/auth';

const AuthContext = createContext({});

export const useAuth = () => {
    const authContext = useContext(AuthContext);
    if (!authContext) {
        throw new Error('Authentication context not found');
    }
    return authContext;
}
const AuthProvider = ({children}) => {

    const [user, setUser] = useState(null);
    const [loading, setLoading] = useState(false);

    const actions = useMemo(
        () => ({
            login: async (email, password) => {
                setLoading(true);
                auth().signInWithEmailAndPassword(email, password)
                    .then((userCredential) => {
                        const currentUser = userCredential.user;
                        console.log('User account signed in!');
                    })
                    .catch(error => {
                        if (error.code === 'auth/user-not-found') {
                            console.log("This email is not yet registered.");
                        }
                        if (error.code === 'auth/wrong-password') {
                            console.log("Invalid password.");
                        }
                        console.log(error);
                    })
                    .finally(() => setLoading(false));
            },
            register: async (name, email, password) => {
                setLoading(true);

                auth().createUserWithEmailAndPassword(email, password)
                    .then((userCredential) => {

                        auth().currentUser.updateProfile({displayName: name});

                        const currentUser = userCredential.user;

                        console.log('User account created!');
                    })
                    .catch(error => {
                        if (error.code === 'auth/email-already-in-use') {
                            console.log('That email address is already in use!');
                        }
                        if (error.code === 'auth/invalid-email') {
                            console.log('That email address is invalid!');
                        }
                        console.error(error);
                    })
                    .finally(() => setLoading(true))
            },
            logout: async () => {
                auth()
                    .signOut()
                    .then(() => {
                        const currentUser = null;
                        console.log('User signed out!');
                    });
            }
        }),
        []
    )

    return (
        <AuthContext.Provider value={{
            user,
            setUser,
            loading,
            setLoading,
            actions
        }}>
            {children}
        </AuthContext.Provider>
    );
};

export default AuthProvider;