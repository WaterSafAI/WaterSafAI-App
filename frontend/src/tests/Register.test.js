import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Register from '../screens/Register';
import { useAuth } from "../services/AuthProvider";

// Mock the useAuth hook
jest.mock("../services/AuthProvider", () => ({
  useAuth: jest.fn(),
}));

describe('Register Component', () => {
  const mockNavigation = { push: jest.fn() };
  const mockRegister = jest.fn();
  beforeEach(() => {
    // Setup mock implementation
    useAuth.mockImplementation(() => ({
      actions: {
        register: mockRegister,
      },
      loading: false,
    }));
  });

  it('renders correctly', () => {
    const { getByText, getByPlaceholderText } = render(<Register />);

    expect(getByText('Create Account')).toBeTruthy();
    expect(getByPlaceholderText('Name')).toBeTruthy();
    expect(getByPlaceholderText('Email')).toBeTruthy();
    expect(getByPlaceholderText('Password')).toBeTruthy();

    const tree = render(<Register navigation={mockNavigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls the register function with correct credentials on button press', () => {
    const { getByPlaceholderText, getByText } = render(<Register />);
    const nameInput = getByPlaceholderText('Name');
    const emailInput = getByPlaceholderText('Email');
    const passwordInput = getByPlaceholderText('Password');
    const createButton = getByText('CREATE');

    fireEvent.changeText(nameInput, 'John Doe');
    fireEvent.changeText(emailInput, 'john@example.com');
    fireEvent.changeText(passwordInput, 'password123');
    fireEvent.press(createButton);

    expect(mockRegister).toHaveBeenCalledWith('John Doe', 'john@example.com', 'password123', 'personal');
  });

  // New Test for Plan Button Press
  it('changes plan state on pressing plan buttons', () => {
    const { getByText } = render(<Register />);
    const personalPlanButton = getByText('Personal');
    const professionalPlanButton = getByText('Professional');

    // Simulate pressing the professional plan button
    fireEvent.press(professionalPlanButton);

    // Check if the register function is called with the 'professional' plan
    const createButton = getByText('CREATE');
    fireEvent.press(createButton);
    expect(mockRegister).toHaveBeenCalledWith(expect.any(String), expect.any(String), expect.any(String), 'professional');
  });
});
