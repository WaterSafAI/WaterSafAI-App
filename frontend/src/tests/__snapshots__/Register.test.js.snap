// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ADD8E6",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-start",
    }
  }
>
  <Text
    style={
      {
        "color": "#0A3465",
        "fontSize": 32,
        "fontWeight": "bold",
        "marginTop": 50,
      }
    }
  >
    Create Account
  </Text>
  <View
    style={
      {
        "backgroundColor": "#D8EBF1",
        "borderColor": "#644535",
        "borderRadius": 15,
        "borderWidth": 3,
        "height": 60,
        "justifyContent": "center",
        "margin": 10,
        "padding": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      fontSize={18}
      onChangeText={[Function]}
      placeholder="Name"
      placeholderTextColor="#644535"
      style={
        {
          "height": 60,
        }
      }
    />
  </View>
  <View
    style={
      {
        "backgroundColor": "#D8EBF1",
        "borderColor": "#644535",
        "borderRadius": 15,
        "borderWidth": 3,
        "height": 60,
        "justifyContent": "center",
        "margin": 10,
        "padding": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      fontSize={18}
      onChangeText={[Function]}
      placeholder="Email"
      placeholderTextColor="#644535"
      style={
        {
          "height": 60,
        }
      }
    />
  </View>
  <View
    style={
      {
        "backgroundColor": "#D8EBF1",
        "borderColor": "#644535",
        "borderRadius": 15,
        "borderWidth": 3,
        "height": 60,
        "justifyContent": "center",
        "margin": 10,
        "padding": 10,
        "width": "80%",
      }
    }
  >
    <TextInput
      fontSize={18}
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="#644535"
      secureTextEntry={true}
      style={
        {
          "height": 60,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 45,
        "justifyContent": "space-between",
        "margin": 20,
        "width": "80%",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "#0A3465",
          },
          {
            "borderColor": "#0A3465",
            "borderRadius": 20,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "width": "45%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#ADD8E6",
            },
            {
              "fontSize": 17,
              "fontWeight": "bold",
              "textAlign": "center",
            },
          ]
        }
      >
        Personal
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "backgroundColor": "#ADD8E6",
          },
          {
            "borderColor": "#0A3465",
            "borderRadius": 20,
            "borderWidth": 3,
            "height": 45,
            "justifyContent": "center",
            "width": "45%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#0A3465",
            },
            {
              "fontSize": 17,
              "fontWeight": "bold",
              "textAlign": "center",
            },
          ]
        }
      >
        Professional
      </Text>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "backgroundColor": "#66B3CC",
        },
        {
          "borderRadius": 15,
          "borderWidth": 3,
          "height": 60,
          "justifyContent": "center",
          "padding": 15,
          "width": "80%",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "#FFFFFF",
          "fontSize": 22,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      CREATE
    </Text>
  </View>
</View>
`;
